### PR TITLE
Improve/fix named threads

### PR DESCRIPTION
**Pull request by BlackMark29A**
This PR addresses two issues regarding named threads:

- The main thread was named `cy-main`, however, user code running in the main thread could spawn new threads, which would then inherit the same name, leading to more than one `cy-main` thread. This behavior is probably more confusing than useful and this PR thus no longer gives the main thread a special name.
- On macOS the `pthread_setname_np` function used to set the name of threads has a different signature than on linux. This in and of itself is not an issue, however, the macOS version can only set the name of the calling thread, which is incompatible with the way celerity currently implements named threads. Therefore this PR adds a feature flag `CELERITY_FEATURE_NAMED_THREADS` to disable the feature entirely on macOS.

### CODE REVIEW

**Review by PeterTh — COMMENTED** [review]
I feel like it would make more sense to use the `#if CELERITY_FEATURE_NAMED_THREADS` to define or stub out `set_thread_name`, rather than doing it at every call site. This keeps the general code (where the call sites are) cleaner.

**Comment by BlackMark29A** [comment]
I wanted to not use a feature flag at all and so my first implementation had a similar idea of having a third stub implementation that would just print a warning that the feature was unsupported, however that obviously didn't work for the unit tests and so I added the feature flag.

I like the idea of keeping the call-sites unchanged, so I'll move the feature flag inside the implementation.

**Comment by BlackMark29A** [comment]
I did not know feature detection macros were intended to be public, in that case if we decide to keep them for this use case we should clearly mark them as being `INTERNAL` or `PRIVATE` or `DETAIL`.

As @PeterTh already pointed out it is cleaner to keep the call-sites unchanged, which I agree with. As for using `__APPLE__` vs a feature detection macro, I would say it depends. Inside `named_threads.unix.cc` I would actually prefer `__APPLE__`, as this relates to platform specific code, however for disabling the unit test I think using `CELERITY_(INTERNAL)_FEATURE_NAMED_THREADS` is better, because the unit test doesn't use any platform specific code, and so disabling it based on feature availability makes more sense to me, especially since the decision whether the feature is available or not is actually made by the build system, and not by the unit test implementation.

Thoughts?

**Comment by fknorr** [comment]
There is precedence for `CELERITY_DETAIL_` in macros.

**Review by fknorr — APPROVED** [review]
LGTM!

File: cmake/celerity-config.cmake.in [hunk 1]
```diff
@@ -50,6 +50,7 @@ set(CELERITY_FEATURE_SCALAR_REDUCTIONS "@CELERITY_FEATURE_SCALAR_REDUCTIONS@")
 set(CELERITY_FEATURE_SIMPLE_SCALAR_REDUCTIONS "@CELERITY_FEATURE_SIMPLE_SCALAR_REDUCTIONS@")
 set(CELERITY_FEATURE_LOCAL_ACCESSOR "@CELERITY_FEATURE_LOCAL_ACCESSOR@")
 set(CELERITY_FEATURE_UNNAMED_KERNELS "@CELERITY_FEATURE_UNNAMED_KERNELS@")
+set(CELERITY_DETAIL_HAS_NAMED_THREADS "@CELERITY_DETAIL_HAS_NAMED_THREADS@")
```
Comment (psalz) [inline cmake/celerity-config.cmake.in:53]: For `_DETAIL_` flags the CMake variable shouldn't be included in the installed package; the preprocessor definition will still be set on anything that links to the runtime library.

**Review by psalz — APPROVED** [review]
One minor remark, otherwise LGTM! Please squash before merging.

File: cmake/celerity-config.cmake.in (hunk 1 above)
Comment (BlackMark29A) [reply cmake/celerity-config.cmake.in:53]: I wasn't sure whether to include it here or not. After discussing it with @fknorr I tentatively settled on including it. But if you say it's not necessary I will remove it.

Comment (BlackMark29A) [reply cmake/celerity-config.cmake.in]: Done.

**Pull request merged**